Android/app/src/main/java/com/peer1/internetmap/NodePopup.java

```diff
@@ -95,4 +167,405 @@ public int getMeasuredHeight() {
+        }
+
+        if (unprocessedHops.size() == 0) {
+            return;
+        }
+
+        isProccessingHop = true;
+
+        Pair<Integer, ProbeWrapper> nextHop = unprocessedHops.remove(0);
+        final Integer ttl = nextHop.first;
+        final ProbeWrapper hopProbe = nextHop.second;
+
+        // Add entry to result list and bump IP hop if necessary
+        // Note: (null, null) indicates a full tracet timeout.
+        if (ttl == null) {
+            addTraceText("Traceroute complete with as many hops as we could contact.");
+            isProccessingHop = false;
+            processNextHop();
+            return;
+        } else if (hopProbe == null) {
+            addTTLResult(ttl, "* * * Hop did not reply or timed out");
+            isProccessingHop = false;
+            processNextHop();
+            return;
+        }
+
+        incrementIPHop();
+
+        // Check to see if we have jumped an ASN
+        CommonClient.getInstance().getApi().getASNFromIP(hopProbe.fromAddress).enqueue(new CommonCallback<ASN>() {
+            @Override
+            public void onRequestResponse(Call<ASN> call, Response<ASN> response) {
+                String asn = response.body().getASNString();
+                NodeWrapper node = null;
+                boolean hasHoppedASN = false;
+
+                if (asn != null) {
+                    node = mapController.nodeByAsn(asn);
+                }
+
+                if (node != null) {
+                    hopNodeWrappers.add(node);
+                    hasHoppedASN = (lastASNIndex != node.index);
+                    lastASNIndex = node.index;
+                }
+
+                if (hasHoppedASN) {
+                    // Update map
+                    incrementASNHop();
+                    displayHops(hopNodeWrappers);
+                }
+
+                addTTLResult(ttl, hopProbe, asn);
+                isProccessingHop = false;
+                processNextHop();
+            }
+
+            @Override
+            public void onRequestFailure(Call<ASN> call, Throwable t) {
+                // TODO what to do if failed to get ASN?
+                isProccessingHop = false;
+                processNextHop();
+                Timber.e("getASNFromIP FAILED YO");
+            }
+        });
+    }
+
+    private void displayHops(ArrayList<NodeWrapper> hops) {
+
+        // If we are no longer running a trace, then do not display hops
+        // Could happen if our request for IPtoASN comes back late.
+        if (tracerouteUtil == null) {
+            return;
+        }
+
+        // Only draw if we can at least make one line.
+        if (hops.size() < 2) {
+            return;
+        }
+
+        // Ew, make this better, maybe pass in cleaner data.
+        NodeWrapper[] mergedHops = hops.toArray(new NodeWrapper[hops.size()]);
+
+        mapController.highlightRoute(mergedHops, mergedHops.length);
+
+
+
+//        NSMutableArray* mergedAsnHops = [NSMutableArray new];
```

Comment:
I had kept it in there as a reference for what iOS was doing, since I am still not 100% positive I am not missing something that chunk of code was doing on Android. 

I will have another look and see if I can determine what the ObjC code was doing here.